src/entry_stream.rs

```diff
@@ -6,87 +6,248 @@ use crate::entry::Entry;
 use std::io::prelude::*;
 use std::net::Shutdown;
 use std::os::unix::net::UnixStream;
 use std::path::Path;
 use std::sync::{Arc, RwLock};
 
-pub trait EntryStreamHandler {
-    fn stream_entries(&mut self, entries: &[Entry]) -> Result<()>;
+pub trait Output: std::fmt::Debug {
+    fn write(&self, payload: String) -> Result<()>;
 }
 
-pub struct EntryStream {
-    pub socket: String,
-    leader_scheduler: Arc<RwLock<LeaderScheduler>>,
+#[derive(Debug, Default)]
+pub struct VecOutput {
+    values: RefCell<Vec<String>>,
 }
 
-impl EntryStream {
-    pub fn new(socket: String, leader_scheduler: Arc<RwLock<LeaderScheduler>>) -> Self {
-        EntryStream {
-            socket,
-            leader_scheduler,
+impl Output for VecOutput {
+    fn write(&self, payload: String) -> Result<()> {
+        self.values.borrow_mut().push(payload);
+        Ok(())
+    }
+}
+
+impl VecOutput {
+    pub fn new() -> Self {
+        VecOutput {
+            values: RefCell::new(Vec::new()),
         }
     }
+
+    pub fn entries(&self) -> Vec<String> {
+        self.values.borrow().clone()
+    }
 }
 
-impl EntryStreamHandler for EntryStream {
-    fn stream_entries(&mut self, entries: &[Entry]) -> Result<()> {
+#[derive(Debug)]
+pub struct SocketOutput {
+    socket: String,
+}
+
+impl Output for SocketOutput {
+    fn write(&self, payload: String) -> Result<()> {
         let mut socket = UnixStream::connect(Path::new(&self.socket))?;
-        for entry in entries {
-            let json = serde_json::to_string(&entry)?;
-            let (slot, slot_leader) = {
-                let leader_scheduler = self.leader_scheduler.read().unwrap();
-                let slot = leader_scheduler.tick_height_to_slot(entry.tick_height);
-                (slot, leader_scheduler.get_leader_for_slot(slot))
-            };
-            let payload = format!(
-                r#"{{"dt":"{}","t":"entry","s":{},"leader_id":"{:?}","entry":{}}}"#,
-                Utc::now().to_rfc3339_opts(SecondsFormat::Nanos, true),
-                slot,
-                slot_leader,
-                json
-            );
-            socket.write_all(payload.as_bytes())?;
-        }
+        socket.write_all(payload.as_bytes())?;
         socket.shutdown(Shutdown::Write)?;
         Ok(())
     }
 }
 
-pub struct MockEntryStream {
-    pub socket: Vec<String>,
-    leader_scheduler: Arc<RwLock<LeaderScheduler>>,
+pub trait EntryStreamHandler {
+    fn emit_entry_event(&self, entries: &Entry) -> Result<()>;
+    fn emit_block_event(&self, tick_height: u64, last_id: Hash) -> Result<()>;
 }
 
+#[derive(Debug)]
+pub struct EntryStream<T: Output> {
+    pub output: T,
+    pub leader_scheduler: Arc<RwLock<LeaderScheduler>>,
+    pub queued_block: Option<EntryStreamBlock>,
+}
+
+impl<T> EntryStreamHandler for EntryStream<T>
+where
+    T: Output,
+{
+    fn emit_entry_event(&self, entry: &Entry) -> Result<()> {
+        let leader_scheduler = self.leader_scheduler.read().unwrap();
```

Comment:
There's some duplicated code between emit_entry_event and emit_block_event:
```rust
        let leader_scheduler = self.leader_scheduler.read().unwrap();
        let slot = leader_scheduler.tick_height_to_slot(tick_height);
        let leader_id = leader_scheduler
            .get_leader_for_slot(slot)
            .map(|leader| leader.to_string())
            .unwrap_or_else(|| "None".to_string());
```
Let's merge?  The `leader_id` also only needs to be computed when the `slot` changes instead of on every single entry.  Move up to the stage loop and pass the results into these two functions?